---
title: "01-BFALgtseq_test1-exploration"
output: html_notebook
---

13 December 2022

Checking on the first GTseq run with the BFAL loci... 
which ones to keep; which ones to toss; how common is the NNNN problem?


```{r load-packages}
library(tidyverse)
```


```{r load-data}
data <- read_csv("../data/BFAL_trimmed_observed_unfiltered_haplotype.csv") %>%
  select(-X1)
```

There are only 75 loci that appear in that dataset. There should be ~350.


```{r}
data %>%
  group_by(locus) %>%
  summarise(sum(depth))
```
Many (most!) entries have only a single read!


A few general assessments:
1) total depth

```{r}
data %>%
  summarise(total_depth = sum(depth))
```


2) depth across loci

```{r read-depth-across-loci}
data %>%
  group_by(locus) %>%
  summarise(total_depth = sum(depth)) %>%
  arrange(desc(total_depth)) %>%
  ggplot(aes(x = reorder(locus, total_depth), y = total_depth)) +
  geom_point()

```

```{r read-depth-across-high-performers}
data %>%
  group_by(locus) %>%
  summarise(total_depth = sum(depth)) %>%
  arrange(desc(total_depth)) %>%
  filter(total_depth > 1000) %>%
  ggplot(aes(x = reorder(locus, total_depth), y = total_depth)) +
  geom_point() +
  theme(
    axis.text.x = element_text(angle = 90)
  )

```
I'll remove the the top 9 overamplifying loci:

scaffold_364_434127
scaffold_925_195725


What about reads across indivs

```{r}
data %>%
  group_by(indiv.ID) %>%
  summarise(total_depth = sum(depth)) %>%
  arrange(desc(total_depth))


```



Make a heatmap of reads/locus/indiv
```{r}
data %>%
  ggplot(aes(x = reorder(locus, depth), y = reorder(indiv.ID, depth), fill = log10(depth))) +
  geom_tile()

```
What if I remove the top loci?

```{r}
data %>%
  filter(!locus %in% c("scaffold_364_434127","scaffold_925_195725") &
           !indiv.ID %in% c("s406", "s434", "s402")) %>%
  ggplot(aes(x = reorder(locus, depth), y = reorder(indiv.ID, depth), fill = log10(depth))) +
  geom_tile() 
```

Loci...

How many actual haplotypes with N base-calls?

```{r}
reads_w_Ns <- data %>%
  filter(!locus %in% c("scaffold_364_434127","scaffold_925_195725") &
   !indiv.ID %in% c("s406", "s434", "s402"),
    str_detect(haplo, "N"),
   depth > 10)

# unique haplotypes with Ns by locus
reads_w_Ns %>%
  select(locus, haplo) %>%
  unique() %>%
  group_by(locus) %>%
  tally() %>%
  arrange(desc(n))

```
We dealt with the N issue by trimming the fasta file (so that there wasn't flanking sequence).

Which loci have only haplotypes with Ns?
```{r}
N_haplos <- data %>%
  mutate(Ns = ifelse(str_detect(haplo, "N"), 1, 0)) %>%
   filter(!locus %in% c("scaffold_364_434127","scaffold_925_195725") &
   !indiv.ID %in% c("s406", "s434", "s402"),
   depth > 2) %>%
  select(locus, haplo, Ns) %>%
  unique() %>%
  group_by(locus) %>%
  mutate(n_haplo = sum(Ns)) %>%
  select(locus, n_haplo) %>%
  unique() # this is the number of haplotypes with Ns by locus

  
```

But to get the ratio, I need additional data.
```{r}
# total number of haplotypes per locus
haplo_per_loc <- data %>%
  mutate(Ns = ifelse(str_detect(haplo, "N"), 1, 0)) %>%
   filter(!locus %in% c("scaffold_364_434127","scaffold_925_195725") &
   !indiv.ID %in% c("s406", "s434", "s402"),
   depth > 2) %>%
  select(locus, haplo) %>%
  unique() %>%
  group_by(locus) %>%
  tally()
  
```

Combine those two things:

```{r}
N_haplos %>%
  left_join(., haplo_per_loc) %>%
  mutate(N_ratio = n_haplo/n) %>%
  filter(N_ratio < 0.5)
```

37 loci in which all haplotypes have at least 1 N
182 loci in which no haplotypes have an N
225 loci in which < 50% of haplotypes have an N


I don't love it, but I'm going to move forward and modify the primer concentrations for the next test run.



NOTE: didn't do this.

Modify primer concentrations for round 2.
```{r}
# primers_test2 <- data %>%
#   group_by(locus) %>%
#   summarise(total_depth = sum(depth)) %>%
#   arrange(desc(total_depth)) %>%
#   mutate(action = ifelse(total_depth > 200000, 0, 1)) %>%
#   mutate(action = ifelse(total_depth > 100 & total_depth < 1000, 1, action)) %>%
#   mutate(action = ifelse(total_depth < 100 & total_depth > 10, 1.5, action)) %>%
#   mutate(action = ifelse(total_depth > 1000 & total_depth < 200000, 0.5, action)) %>%
#   mutate(action = ifelse(total_depth < 10, 1.75, action)) %>%
#   # add in the original primer concentration
#   mutate(primer_conc = 0.25, new_conc = action*primer_conc, stock_conc = 100, pool_vol = 1000, cherrypick_vol = (new_conc*pool_vol)/stock_conc) %>%
#   # now split the pool into 2 based on read depth
#   mutate(primer_pool = ifelse(total_depth > 200, "Pool1", "Pool2")) %>%
#   group_by(primer_pool) %>%
#   add_tally() #%>%
#   #write_csv("../outputs/BFALgtseq_test2_primer_pool.csv")


```


### Primer pool for round 2
3 January 2022

Read in the overampers file from the `microhaplot_local.R` script.
```{r}
overampers <- read_csv("../results/test1_overampers.csv")

# fix the formatting
overampers1 <- gsub("_PCR_Product", "", overampers$locus)

overampers$locus <- overampers1

overampers
```



One more thing needed for the cherrypick file: in which well is each primer located?

```{r}
# read in primer plate map
platemap <- readxl::read_xlsx("../IDTplateSpecs_BFALgtseq_20221012.xlsx")
```


```{r}
# select the relevant variables and make the sequence name match the data
platemap1 <- platemap %>%
  select(`Plate Name`, `Well Position`, `Sequence Name`) %>%
  rename(locus = `Sequence Name`) 


tmp <- gsub("BFAL_", "", platemap1$locus)
tmp1 <- gsub("_R", "", tmp)
tmp2 <- gsub("_F", "", tmp1)
  
platemap1$locus <- tmp2

platemap1
```


```{r}
# merge those two things:
# platemap1 %>%
#   left_join(., primers_test2) %>%
#   write_csv("../outputs/BFALgtseq_test2_primer_pool.csv")

overampers %>%
  left_join(., platemap1) %>%
  write_csv("../outputs/BFALgtseq_test2_primer_to_remove.csv")
```



## Taking a look at output from GTscore

The plot we want is the R1 reads where the FWD primer was found vs. the flashed read where the FWD primer was found.


```{r}
# read in the GTscore output

flashed <- read_tsv("../data/flashed_GTscore_locusSummary.txt") %>%
  select(Locus, `Primer Reads`) %>%
  rename(flashed_reads = `Primer Reads`)

R1 <- read_tsv("../data/R1_GTscore_locusSummary.txt") %>%
  select(Locus, `Primer Reads`) %>%
  rename(R1_reads = `Primer Reads`)
```

Look at those two df combined based on locus
```{r}
combo_df <- R1 %>%
  left_join(., flashed, by = "Locus")

combo_df %>%
  summarise(sum(R1_reads))

combo_df %>%
  summarise(sum(flashed_reads))
```


```{r}
combo_df %>%
  ggplot(aes(x = flashed_reads, y = R1_reads)) +
  geom_point() +
  labs(title = "BFAL flashed vs. unmerged reads") +
  theme_minimal()

ggsave("pdf_outputs/BFAL_gtseq_test1_offtarget.pdf")

```

So the loci that would be concerning are those with high numbers for the unmerged (R1) reads and disproportionately low numbers for the flashed reads in comparison.

```{r}
# create a proportion difference
combo_df %>%
  mutate(prop = flashed_reads/R1_reads) %>%
  ggplot(aes(x = prop)) +
  geom_histogram() +
  labs(x = "proportion of R1 that also merged",
       title = "BFAL proportion of on-target reads") +
  theme_minimal()

ggsave("pdf_outputs/prop_ontarget.pdf")
```

Look at the loci with < 50% on-target (because read depths were low for this run, I'm not necessarily going to drop these)


```{r}
combo_df %>%
  mutate(prop = flashed_reads/R1_reads) %>%
  filter(prop < 0.5,
         R1_reads > 10000) %>%
  ggplot(aes(x = prop)) +
  geom_histogram() +
  labs(x = "proportion of R1 that also merged")


```
For these loci, I think I'm only concerned if there were legitimately a lot of reads in the R1 reads. Otherwise, it's probably just an issue of low depth.


Seems like I could easily draw the line at 0.05.

```{r}
offtarget_locs <- combo_df %>%
  mutate(prop = flashed_reads/R1_reads) %>%
  filter(prop < 0.05,
         R1_reads > 10000)

offtarget_locs
```
10 loci to potentially drop.

Are these distinct from the loci to remove based on the idx stats?

```{r}
# fix the formatting
tmp <- gsub("BFAL_", "", offtarget_locs$Locus)
offtarget_locs$Locus <- tmp

offtarget.df <- offtarget_locs %>%
  mutate(drop = "off-target") %>%
  rename(locus = Locus)

locs_to_lose <- overampers %>%
  mutate(drop = "overamper") %>%
  full_join(., offtarget.df) %>%
  arrange(locus)


locs_to_lose
```


```{r}
locs_to_lose %>%
  left_join(., platemap1) %>%
  select(locus, drop, `Plate Name`, `Well Position`) %>%
  write_csv("../outputs/BFAL_primers_to_remove_20230104.csv")

```

