---
title: "10-locus-summary-and-eval"
output: html_notebook
---

23 February 2023

Evaluating the loci that we're using for self-assignment and bycatch assignment.

These were originally derived from lcWGS data and the microhaplotype results suggest that the real variation present is potentially quite different.

Initially, I should probably just focus on the baseline genotypes rather than including everything.


To evaluate the markers, I need to subset the reference baseline genotypes for just the minimal amount of missing data.

```{r load-libraries}
library(tidyverse)
library(adegenet)
library(radiator)
library(DescTools)

```

Once again, I need to finagle the dataset into a genind object.

```{r}
# read in data
reference_genos <- read_rds("../data/processed/reference_genos.rds")

```

Check on missing data:
```{r}
# number of missing alleles (max = 159*2)
# out of 159 loci
inds_to_toss_missing_data <- reference_genos %>%
  group_by(indiv) %>%
  filter(is.na(alleidx)) %>%
  group_by(indiv, collection) %>%
  tally() %>%
  arrange(desc(n)) %>%
  filter(n > 9)
  
# if I removed all the indivs with > 10 missing alleles, that's only 5 indivs.
# what would the distribution of indivs per population be? 
# still ok, I'm pretty sure.
reference_genos %>%
  anti_join(., inds_to_toss_missing_data) %>%
  group_by(collection, indiv) %>%
  tally() %>%
  tally()

```
I feel fine about that. Let's see if that makes any of the HWE calculations easier because less missing data?

```{r}
ref_genos_low_missing_data <- reference_genos %>%
  anti_join(., inds_to_toss_missing_data)

```



Make the df match the requirements for tidy_genomic_data
```{r}
long_df <- ref_genos_low_missing_data  %>%
  select(-gene_copy) %>%
  select(collection, everything()) %>%
  rename(INDIVIDUALS = indiv, STRATA = collection, MARKERS = locus, GT = alleidx)

```

Genotypes should be coded with 3 integers for each alleles. 6 integers in total for the genotypes. e.g. 001002 or 111333 (for heterozygote individual). 6 integers WITH separator: e.g. 001/002 or 111/333 (for heterozygote individual). The separator can be any of these: "/", ":", "_", "-", ".", and will be removed.


```{r}
# create 3 digit integers from the genotypes
long_df$GT3 <- Format(long_df$GT, ldigits = 3, digits = 0)

# fix NAs
long_df0s <- long_df %>%
  mutate(GT3 = ifelse(is.na(GT3), "000", GT3)) # I don't love that this creates potential artifacts!
```

Now combine the GT3 column per indiv/marker:
```{r}
# make the genos characters and then try pasting them as strings
long_df0s$GT3 <- as.character(long_df0s$GT3)

long_df3digit <- long_df0s %>%
  group_by(INDIVIDUALS, MARKERS) %>% 
  arrange(GT3, .by_group = TRUE) %>% 
  summarise(GENOTYPE = toString(GT3))
  
# paste strings together
long_df3digit$GENOTYPE <- gsub(", ","",long_df3digit$GENOTYPE)


# add back on species identity as strata
df_for_conversion <- long_df0s %>% 
  select(-GT, -GT3) %>%
  left_join(., long_df3digit) %>%
  unique() %>%
  rename(GT = GENOTYPE) %>%
  mutate(GT = ifelse(GT == "000000", NA, GT)) %>%
  filter(!STRATA %in% c("Whale-Skate", "Lehua"))
  

df_for_conversion$STRATA <- as.factor(df_for_conversion$STRATA)

```

Double check - how many samples per population?
```{r}
df_for_conversion %>%
  select(INDIVIDUALS, STRATA) %>%
  unique() %>%
  group_by(STRATA) %>%
  tally()

```
Let's remove Lehua and Whale-Skate at this point because of so few samples.

```{r}
convert_df_wo_2pops <- df_for_conversion 
```



```{r convert-df-to-genind}
# use the radiator package for this conversion
genind_df <- write_genind(convert_df_wo_2pops)

```
Basic population genetic evaluation of the markers, following, in part, this:
https://bookdown.org/hhwagner1/LandGenCourse_book/WE_3.html


```{r}
sum_output <- summary(genind_df)

names(sum_output)

expected_hz <- as.data.frame(sum_output$Hexp)
observed_hz <- as.data.frame(sum_output$Hobs)


```

Expected heterozygosity (here: Hexp) is the heterozygosity expected in a population under HWE, and observed heterozygosity (here: Hobs) is the observed number of heterozygotes at a locus divided by the total number of genotyped individuals. Here are the global values (pooled across all populations):

```{r}
# expected heterozygosity per population
adegenet::Hs(genind2genpop(genind_df))
```
I do wonder if I should combine some of the smaller pops - e.g., Whale-Skate, which is part of the French Frigate Shoals (FFS), which includes Tern Island.

For now, I can remove Whale-Skate and discuss with Jessie if we want to make that a single reporting group.


```{r}
Hobs <- t(sapply(seppop(genind_df), function(ls) summary(ls)$Hobs))
  Hexp <- t(sapply(seppop(genind_df), function(ls) summary(ls)$Hexp))
  {cat("Expected heterozygosity (Hexp):", "\n")
  round(Hexp, 2)
  cat("\n", "Observed heterozygosity (Hobs):", "\n")
  round(Hobs, 2)}
  
```


```{r}
par(mar=c(5.5, 4.5, 1, 1))
  Hobs.pop <- apply(Hobs, MARGIN = 1, FUN = mean)
  Hexp.pop <- apply(Hexp, 1, mean) 
  barplot(Hexp.pop, ylim=c(0,1), las=3, ylab="Expected heterozygosity")
  barplot(Hobs.pop, ylim=c(0,1), las=3, ylab="Observed heterozygosity")
```
```{r}
#install.packages("PopGenReport")

# Richness <- PopGenReport::allel.rich(genind_df, min.alleles = NULL)
# Richness$alleles.sampled
```


```{r}
bfal_diversity <- data.frame(Pop = names(Hobs.pop),
                              n_samples = sum_output$n.by.pop,
                              Hobs = Hobs.pop,
                              Hexp = Hexp.pop)
                              #Ar = Richness$mean.richness)
as.tibble(bfal_diversity) %>%
  rename(Population = Pop, H_obs = Hobs, H_exp = Hexp) #%>%
  #write_csv("csv_outputs/reference_pop_hz_summary.csv")

```
chi^2: value of the classical chi-squared test statistic
df: degrees of freedom of the chi-squared test
Pr(chi^2 >): p-value of the chi-squared test (‘>’ indicates that the alternative is ‘greater,’ which is always the case for a chi-squared test)
Pr.exact: p-value from an exact test based on Monte Carlo permutation of alleles (for diploids only). The default is B = 1000 permutations (set B = 0 to skip this test). Here we use the function ‘round’ with argument ‘digits = 3’ to round all values to 3 decimals.

https://rdrr.io/cran/pegas/man/hw.test.html

In this case, the degrees of freedom depend on the number alleles for a given locus.

```{r}
# HWE test for all loci - but really what we want is all loci in each population (farther below)
hwe_output <- round(pegas::hw.test(genind_df, B = 1000), digits = 3)

# which ones are statistically sign.
rownames_to_column(data.frame(hwe_output)) %>%
  filter(Pr.exact < 0.05)

```
26 loci that are out of HWE globally. (included is the funky one, scaffold_75_1184494) [31 with the missing data indivs removed from the baseline]

After looking at the loci x population HWE info, I can remove any that are out of HWE in the majority of populations.


```{r}
# Chi-squared test: p-value
HWE.test <- data.frame(sapply(seppop(genind_df), 
                              function(ls) pegas::hw.test(ls, B=0)[,3]))
HWE.test.chisq <- t(data.matrix(HWE.test))
{cat("Chi-squared test (p-values):", "\n")
round(HWE.test.chisq,3)}

# these are the p-values for the per-pop calcs
loci_to_toss_hwe <- rownames_to_column(as.data.frame(HWE.test.chisq), var = "population") %>%
  pivot_longer(cols= 2:length(rownames_to_column(as.data.frame(HWE.test.chisq))), names_to = "locus", values_to = "p_val") %>%
  filter(p_val < 0.05) %>% # which p-values are significant?
  group_by( locus) %>%
  tally() %>% # how many loci are out of HWE in how many populations?
  arrange(desc(n)) %>%
  filter(n>2)

```

Remove loci that are statistically out of HWE in >2 populations (the majority).


That's only 5 loci.

```{r}
loci_to_toss_hwe %>%
  write_csv("csv_outputs/loci_to_toss_hwe.csv")

# from here, I can generate a "final" dataset.
```

Pivoting to a new Rmd `11-hwe-filtered-baseline-and-mixture.Rmd`.



```{r}
# Monte Carlo: p-value
HWE.test <- data.frame(sapply(seppop(genind_df), 
                              function(ls) pegas::hw.test(ls, B=1000)[,4]))
HWE.test.MC <- t(data.matrix(HWE.test))
{cat("MC permuation test (p-values):", "\n")
round(HWE.test.MC,3)}
```


```{r}
alpha=0.05
Prop.loci.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 2, mean), 
           MC=apply(HWE.test.MC<alpha, 2, mean))

Prop.loci.out.of.HWE
# seems like the MC is less conservative, but when MC is 0.2, Chisq is at least that

Prop.loci.out.of.HWE %>%
  filter(MC > 0) # 19 loci according to MC; actually 2 instances where Chisq doesn't pick it up

```

```{r}
data.frame(Chisq=apply(HWE.test.chisq<alpha, 1, mean))

data.frame(HWE.test.chisq)
```


```{r}
Prop.pops.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 1, mean), 
           MC=apply(HWE.test.MC<alpha, 1, mean))

Prop.pops.out.of.HWE   
```
Check what's going on here (above).


```{r}
Chisq.fdr <- matrix(p.adjust(HWE.test.chisq,method="fdr"), 
                    nrow=nrow(HWE.test.chisq))
MC.fdr <- matrix(p.adjust(HWE.test.MC, method="fdr"), 
                    nrow=nrow(HWE.test.MC))

Prop.pops.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 1, mean), 
           MC=apply(HWE.test.MC<alpha, 1, mean),
           Chisq.fdr=apply(Chisq.fdr<alpha, 1, mean),
           MC.fdr=apply(MC.fdr<alpha, 1, mean))
Prop.pops.out.of.HWE             
```

Correct for the number of tests:
```{r}
Chisq.fdr <- matrix(p.adjust(HWE.test.chisq,method="fdr"), 
                    nrow=nrow(HWE.test.chisq))
MC.fdr <- matrix(p.adjust(HWE.test.MC, method="fdr"), 
                    nrow=nrow(HWE.test.MC))

Prop.pops.out.of.HWE <- data.frame(Chisq=apply(HWE.test.chisq<alpha, 1, mean), 
           MC=apply(HWE.test.MC<alpha, 1, mean),
           Chisq.fdr=apply(Chisq.fdr<alpha, 1, mean),
           MC.fdr=apply(MC.fdr<alpha, 1, mean))
Prop.pops.out.of.HWE             
```







### PCA


Now that the data is a genind object, go ahead and run the PCA.

Make PCA
```{r dataset-pca}
# Allele presence absence data are extracted and NAs replaced using tab:
datasetX <- tab(genind_df, NA.method="mean") # double check that is this the appropriate method.

# make PCA
dataset_pca1 <- dudi.pca(datasetX, center = TRUE, scannf = FALSE, scale=FALSE, nf = 1000)

# colors
mycol <- colorRampPalette(c("deepskyblue", "darkgreen", "brown", "midnightblue", "darkcyan", "slateblue", "purple4"))(6)

# plot with factor labels
pdf("pdf_outputs/BFALgtseq_reference_filtered_genos_pca.pdf", width = 10, height = 10)
s.class(dataset_pca1$li, fac=pop(genind_df), wt = rep(1, length(pop(genind_df))), clabel = .8, grid = FALSE, cellipse = 2,
        xax=1, yax=2, col=transp(mycol,.8),
        axesel=FALSE, cstar=0, cpoint=1)
dev.off()

```


Out of curiosity, filter the genind to remove Torishima and then re-run the PCA with just the Hawaiian colonies.

```{r}
# remove Torishima for test PCA
convert_wo_3pops <- convert_df_wo_2pops %>%
  filter(STRATA != "Torishima")

# use the radiator package for this conversion
genind_wo_Torishima <- write_genind(convert_wo_3pops)


# Allele presence absence data are extracted and NAs replaced using tab:
dataset_wo_Tor <- tab(genind_wo_Torishima, NA.method="mean") # double check that is this the appropriate method.

# make PCA
dataset_wo_Tor1 <- dudi.pca(dataset_wo_Tor, center = TRUE, scannf = FALSE, scale=FALSE, nf = 1000)

# colors
mycol <- colorRampPalette(c("deepskyblue", "darkgreen", "brown", "midnightblue", "darkcyan", "slateblue", "purple4"))(6)

# plot with factor labels
pdf("pdf_outputs/BFALgtseq_reference_filtered_genos_pca_wo_Torishisma.pdf", width = 10, height = 10)
s.class(dataset_wo_Tor1$li, fac=pop(genind_wo_Torishima), wt = rep(1, length(pop(genind_wo_Torishima))), clabel = .8, grid = FALSE, cellipse = 2,
        xax=1, yax=2, col=transp(mycol,.8),
        axesel=FALSE, cstar=0, cpoint=1)
dev.off()
```



## Check on a few of the multi-hap loci from test 5

These looks funky: 
scaffold_75_1184494
scaffold_7_15896

These are also two loci that came up as out of HWE. Perfect. I'll remove them and the other loci that had HWE issues across the majority of pops.


```{r}
reference_genos %>%
  filter(locus %in% c("scaffold_7_15896", "scaffold_75_1184494"))


```


